File: src/main/java/pl/training/workshop/m5/s15_compatibility/step3/TicketExporter.java
package pl.training.workshop.m5.s15_compatibility.step3;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.Comparator;

import pl.training.workshop.m5.s15_compatibility.Column;

/**
 * Krok 3: bez zmian.
 */
public final class TicketExporter {
    public String export(Ticket ticket) {
        StringBuilder row = new StringBuilder(ticket.title());
        Method[] candidates = ticket.getClass().getMethods();
        Arrays.stream(candidates)
                .filter(method -> method.isAnnotationPresent(Column.class))
                .sorted(Comparator.comparing((Method m) -> m.getAnnotation(Column.class).value()))
                .forEach(method -> row.append(';')
                        .append(method.getAnnotation(Column.class).value())
                        .append('=')
                        .append(read(method, ticket)));
        return row.toString();
    }

    private static Object read(Method method, Ticket ticket) {
        try {
            return method.invoke(ticket);
        } catch (IllegalAccessException | InvocationTargetException e) {
            throw new IllegalStateException(e);
        }
    }
}

File: csharp/src/Training.Workshop/M5/S15Compatibility/Step1/TicketExporter.cs
using System.Reflection;
using System.Text;
using Training.Workshop.M5.S15Compatibility;

namespace Training.Workshop.M5.S15Compatibility.Step1;

/// <summary>
/// Krok 1: przygotowanie refleksji PRZED ruchem w hierarchii - <c>GetMethods()</c> widzi też
/// publiczne metody odziedziczone. Wynik bez zmian, ale kod jest gotowy na Pull Up.
/// </summary>
public sealed class TicketExporter
{
    public string Export(Ticket ticket)
    {
        var row = new StringBuilder(ticket.Title);
        var candidates = ticket.GetType().GetMethods();
        foreach (var method in candidates
                     .Where(method => method.IsDefined(typeof(ColumnAttribute)))
                     .OrderBy(method => method.GetCustomAttribute<ColumnAttribute>()!.Value, StringComparer.Ordinal))
        {
            row.Append(';')
                .Append(method.GetCustomAttribute<ColumnAttribute>()!.Value)
                .Append('=')
                .Append(Read(method, ticket));
        }
        return row.ToString();
    }

    private static object? Read(MethodInfo method, Ticket ticket)
    {
        try
        {
            return method.Invoke(ticket, null);
        }
        catch (TargetInvocationException e)
        {
            throw new InvalidOperationException(e.Message, e);
        }
    }
}
